### DISCUSSION TITLE
How to use custom session store

### DISCUSSION THREAD

**Discussion by JaceEgg** [jacobobryant/biff#215] [General]
Does anyone have an example of how to inject your own implementation of ring.middleware.session.store? I wanted to try implementing one using XTDB, but couldn't quite figure out how to wire it up without copy/pasting and altering `wrap-session`: https://github.com/jacobobryant/biff/blob/edfd869b5866503d0ca2414874c1f37607c268ef/src/com/biffweb/impl/middleware.clj#L115

**Comment by jacobobryant** [comment]
First you'll need to remove this line from your config.edn file: https://github.com/jacobobryant/biff/blob/edfd869b5866503d0ca2414874c1f37607c268ef/starter/resources/config.edn#L18

Then you can create a `use-xtdb-session-store` biff component that adds a `:biff.middleware/session-store` key to the system map. It'll need to go after the use-xtdb component (https://github.com/jacobobryant/biff/blob/edfd869b5866503d0ca2414874c1f37607c268ef/starter/src/com/example.clj#L66) but before the use-jetty component. use-jetty merges the system map into incoming requests, so that's how wrap-session gets access to it.

**Reply by jacobobryant** [reply]
The component would look something like this:

```clojure
(defn use-xtdb-session-store [{:keys [biff.xtdb/node] :as ctx}]
  (assoc ctx :biff.middleware/session-store ...))
```

By the way, when you have that working, if you stick the code in a gist or something I'll put a link to it in the content library. I bet there are others who would be interested in that.

**Reply by JaceEgg** [reply]
Thanks for the instructions. Here is what I came up with: https://gist.github.com/JaceEgg/a4150b8f360110d1cb56a5744aeca4d4
Feel free to make changes to it if you put it in the content library, I am not very familiar with clojure.

**Reply by jacobobryant** [reply]
Looks great! I just added a link [here](https://biffweb.com/docs/library/).
